primitives/state-machine/src/proving_backend.rs

```diff
@@ -112,15 +115,42 @@ impl<'a, S, H> ProvingBackendRecorder<'a, S, H>
 	}
 }
 
-/// Global proof recorder, act as a layer over a hash db for recording queried
-/// data.
-pub type ProofRecorder<H> = Arc<RwLock<HashMap<<H as Hasher>::Out, Option<DBValue>>>>;
+/// A type that records all accessed trie nodes.
+pub enum ProofRecorder<H: Hasher> {
+	/// Records are separated by child trie, this is needed for
+	/// proof compaction.
+	Full(Arc<RwLock<ChildrenMap<RecordMapTrieNodes<H>>>>),
+	/// Single storage for all recoded nodes (as in
+	/// state db column).
+	/// That this variant exists only for performance
+	/// (on less map access than in `Full`), but is not strictly
```

Comment:
Why not store the `ChildInfo` alongside the values there? You can have an `Arc` or `Rc` there and it would still require only one lookup and wouldn't really consume more memory.

Reply:
For the case where we use the Full variant we would need to split the content by child info afterward, I have no idea if it will be faster or not. Generally (as metioned in the PR description) I think this enum is not really needed, will see during refactoring if it can be dropped.